torch_geometric/contrib/nn/models/hin2vec.py

```diff
@@ -0,0 +1,375 @@
+            self.walks_per_node)
+        rw = self.random_walk_fn(self.rowptr, self.col, start_nodes,
+                                 self.walk_length, 1.0, 1.0)
+        node_seq, edge_seq = rw
+
+        # -1 indicates no edge in random walk
+        mask = edge_seq < 0
+        edge_seq = edge_seq.clamp(min=0, max=len(self.edge_id) - 1)
+
+        edge_id_seq = self.edge_id[edge_seq]
+        edge_id_seq[mask] = -1
+        return node_seq, edge_id_seq
+
+    def _extract_distinct_metapath(self, rw: Tensor) -> Dict[int, Tensor]:
+        r"""Extract distinct metapath within self.metapath_length hops from
+        each random walk. Each metapath is a sequence of edge types.
+        """
+        distinct_path_dict = {}
+        for w in range(self.metapath_length):
+            metapath_w = []
+            for i in range(rw.size(1) - w):
+                metapath = rw[:, i:i + w + 1]
+                # Filter out row with -1 value which indicates end of random
+                # walk due to no next edge.
+                mask = metapath >= 0
+                mask = mask.all(1)
+                metapath = metapath[mask, :]
+                metapath = torch.unique(metapath, dim=0)
+                metapath_w.append(metapath)
+
+            unique_metapath_w = torch.unique(torch.cat(metapath_w), dim=0)
+
+            distinct_path_dict[w + 1] = unique_metapath_w
+        return distinct_path_dict
+
+    def test(
+        self,
+        train_z: Tensor,
+        train_y: Tensor,
+        test_z: Tensor,
+        test_y: Tensor,
+        solver: str = 'lbfgs',
+        *args,
+        **kwargs,
+    ) -> float:
+        r"""Evaluates latent space quality via a logistic regression downstream
+        task.
+        """
+        from sklearn.linear_model import LogisticRegression
+
+        clf = LogisticRegression(solver=solver, *args,
+                                 **kwargs).fit(train_z.detach().cpu().numpy(),
+                                               train_y.detach().cpu().numpy())
+        return clf.score(test_z.detach().cpu().numpy(),
+                         test_y.detach().cpu().numpy())
+
+    def __repr__(self) -> str:
+        return (f'{self.__class__.__name__}('
+                f'{self.node_embedding.weight.size(0)}, '
+                f'{self.node_embedding.weight.size(1)})')
+
+
+class _BinaryStepFunction(Function):
+    @staticmethod
+    def forward(input: Tensor) -> Tensor:
+        return (input > 0).float()
+
+    @staticmethod
+    def setup_context(ctx: Any, inputs: Any, output: Any):
+        input, = inputs
+        ctx.save_for_backward(input)
+
+    @staticmethod
+    def backward(ctx: Any, grad_output: Tensor) -> Tensor:
+        input, = ctx.saved_tensors
+        dx = input.clone()
+        mask = (-6 <= dx) & (dx <= 6)
+        dx[mask] = torch.sigmoid(dx[mask]) * (1 - torch.sigmoid(dx[mask]))
+        return grad_output * dx
+
+
+class _BinaryStepModule(torch.nn.Module):
```

Comment:
The unit test error seems related to this. Maybe we can use something like
```python
from packaging import version

TORCH_VERSION = version.parse(torch.__version__)
if TORCH_VERSION < version.parse("...")...
   class ...
else:
   class ...
```
to make sure it's also working for the older version?

Reply:
I figured out that the issue is  actually with `Torch.Autograd.Function` (_BinaryStepFunction) implementation. There are 2 ways to define `forward()` and `setup_context()`, ie. combined or separate. Pytorch recommend [second way ](https://pytorch.org/docs/stable/notes/extending.html#how-to-use), which I implemented. 

Apparently, this is not compatible to pytorch version earlier than 2.0 after some debugging and [checking](https://pytorch.org/docs/stable/notes/extending.html#how-to-use). I updated it to the combined method instead, to be compatible to both earlier and current version. 

Now, all the tests have passed.